Draft: xapp specification (#1)
xapp-project/xapp-project: Specification.md

Reviewer: I would suggest that the specification adheres to [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119) to make it clearer what is mandated, optional, etc.

Author: Personally I am not a big fan, though fine for me, if you think it helps to make things more clear :+1:

Author: Done: 73f808df362596b809b7b6205dc5e6bf99d88add - closing this thread

# XApp Specification

> TBD: I suppose it would be nice to call the project XApp instead of xapp ? (used lower case for now)

Reviewer 2: XApp sounds good.

Author: Fixed: 7ddaa169f8ebfa702d07b975ad6aae174b7f973c  … @clefebvre, can you take care for the renaming of the GitHub project? (I suppose only you have the required permissions for that)

In order to be shown on the xapp project, applications need to fulfill a list of requirements. This specification is used in order to summarize these requirements and as such can be utilized to decide if an application can join the xapp project.

## Requirements

### Licensing

> Xapps need to be licensed on a [FOSS license](https://opensource.org/licenses).

Reviewer: ```suggestion
Xapps need to be licensed on an [OSI-approved license](https://opensource.org/licenses).
```

Is it my understanding that there will not be any mandating of copyleft vs permissive licensing? Only asking for clarification, not advocating for there to be such mandate.

Author: Fixed - at least speaking for myself, I don't see a need for more strict licensing conditions.

> Xapps need to be licensed on an [OSI-approved license](https://opensource.org/licenses).

Reviewer 3: nit: `s/on/under/`

Author: Thanks, fixed: 3b7f11828cfa9c079108e2bce558599c9e477933

### Operating System

> Xapps need to be able to run in a Linux environment. The support of other operating systems is optional.

Reviewer 3: This feels a little bit too negative toward other OSes.  I think it's fine that Linux is the only requirement, but it might be friendlier to include something like, "Support for other operating systems is optional, but application developers should strive to write portable code where possible."

Author: Hmm, not sure for this one. It sounds a bit like xapp devs should know how to write code which is good to port across OSes.

Maybe like this to get the phrase less negative: "Support for other operating systems is welcome, though optional." ?

Author: FIxed: 0f7836e9970d867d1ab141c3f9f055e68fa32fd0

### Graphical Desktop Application

> This project targets graphical applications which run on desktop systems and dependencies of them. E.g. console applications, kernel-modules and system-services cannot be xapps.

Reviewer: If one of the goals (at least based off my understanding from engaging in discussions with @clefebvre and in Matrix)  is for Xapps to feature components used to compose desktop environments and share that engineering work, then console applications and system-services should be permitted within the project. System services could be seen as programs such as XDG portals and daemons which may not directly have graphical components. Console applications for bootstrapping, configuration, DX (developer-related tooling). would also end up not being included.

Author: If you think it makes sense to as well include such application, that's fine for me. I just had the impression the project is supposed to be only about graphical apps, though I might be wrong at this. If it is fine for @clefebvre I suppose we can just drop that section.

Reviewer 2: It's a good question.

xdg-desktop-portal-xapp comes to mind. Is it a graphical app? Not really... Is is an XApp project? Absolutely. Should it be? I think so yes.

If tomorrow we work on a settings daemon or a CLI tool, does it really matter if they have a UI or not? Probably not.

It seems to me that the usage, audience and platform target (who's using it, in which DEs? etc..) matters more than whether there's a UI or not.

Author: Alright, just dropped that section: 2d1e3d46819f6d25a67be47262f5fcf64c673183

### Usage of standards

> In order to be desktop-environment agnostic, the compliance to the following [freedesktop.org](https://www.freedesktop.org) standards is mandatory to be listed as an xapp:

Reviewer 3: I think this requires a little more explanation; to me it reads like all xapps need to comply with these standards, even when they might not need to use them.  Like a calculator app probably won't care about the file manager dbus interface or shared mime database.  Maybe something more like:

"In order to be desktop-environment agnostic, xapps must comply with the following standards when they apply to functionality included in the application."

Author: Agreed, thats a better wording .. fixed in a4c89db39fb2e19979d3818defecb26983ce3b87.

- [Free Media Player Specification](https://www.freedesktop.org/wiki/Specifications/free-media-player-specs/)
- [Free Media Player Remote Interface Specification](https://www.freedesktop.org/wiki/Specifications/mpris-spec/)
- [Xdg Icon Theme](https://www.freedesktop.org/wiki/Specifications/icon-theme-spec/)
- [Xdg Icon Naming](https://www.freedesktop.org/wiki/Specifications/icon-naming-spec/)
- [Shared mime Database](https://www.freedesktop.org/wiki/Specifications/shared-mime-info-spec/)
- [Startup Notification](https://www.freedesktop.org/wiki/Specifications/startup-notification-spec/)
- [Trashcan support](https://www.freedesktop.org/wiki/Specifications/trash-spec/)
- [Bookmark Interchange Format](https://pyxml.sourceforge.net/topics/xbel/)

TBD: Not sure on these:
- [MIME Applications Associations](https://www.freedesktop.org/wiki/Specifications/mime-apps-spec/)
- [XSettings](https://www.freedesktop.org/wiki/Specifications/xsettings-spec/)
- [Cursor Conventions](https://www.freedesktop.org/wiki/Specifications/cursor-spec/)
- [Sound Theme](https://www.freedesktop.org/wiki/Specifications/sound-theme-spec/)

### Window decorations

> Applications which are part of the xapp project should support server side window decorations and use them by default. Xapps are free to optionally as well support client side decorations [^1].

Reviewer: I don't see how this is enforceable. GNOME Wayland notoriously doesn't support server-side decorations and instead applications have to resort to using things like libdecor to generate client-side decorations.

Author: Xapps which want to be able to run in GNOME wayland can still use CSD. Though, the default should be SSD.

An xapp which wants to support GNOME wayland might want to probe if it is running in a GNOME wayland session at startup, and if so, it can use CSD.

Author: Imo the fact that gnome does not support SSD can be considered as a GNOME problem ... personally I don't plan to install optional CSD into Xfce apps, just to please gnome ideas.

Reviewer 3: For Wayland, apps should be responsive to [KDE server decoration](https://wayland.app/protocols/kde-server-decoration)/[XDG decoration](https://wayland.app/protocols/xdg-decoration-unstable-v1), though xapps should request/set SSD mode.  If the compositor insists on CSD (or doesn't support either of these protocols), the app will either have to draw CSDs, or just be fine with the idea that they'll be undecorated on compositors with no SSD support (which is fine).  So I think the wording I'd like to see here is more like:

"Applications which are part of the xapp project should default to server-side decorations via a windowing system-specific protocol, if one is available.  If an xapp is run on a desktop environment that prefers or requires applications to draw client-side decorations, the xapp may do so, but is not required by this specification to support this mode of operation."

My rationale here is that I could see a situation where the desktop environment itself has no strong opinion on SSD vs. CSD, but allows the user to choose what to prefer (e.g. Wayfire has such a setting).  The xapp should obey the user's preference , assuming the xapp's developer wants to include CSD support.

Author: The wording suggested by @kelnos sounds good and reasonable to me. (As someone who has not that much knowledge on the topic)

@JoshStrobl , would you agree with that suggestion?

Reviewer 2: I think the devs might want to support this in some cases but not others. Let me take an example.

**Scenario 1**

Imagine two calculator projects, both XApps. One has a menubar/SSD, one is CSD. The two teams talk to each others but don't feel the need to merge their efforts. These projects are small anyway. DEs appreciate the choice and don't all use the same calculator.

**Scenario 2**

Something that didn't exist before and is hard to maintain comes up. DEs want to use it but they don't all want the exact same UI.

In this scenario we don't want to duplicate the work into multiple projects, the devs might want to provide configurable UI choices.

------------

All in all, I think it's something that's a case by case thing. It depends largely on who wants to use the app, who's making it, whether or not there are other XApp alternatives etc...

I wouldn't make the SSD a requirement. I think what matters the most is the general intention. If a stakeholder wants to use something, it's in the nature of Xapps to accomodate to a certain degree.

Author: Though doesn't an app which only supports CSD collide with the requirement to be freely styleable by the user? IIrc, having user-stylable apps was one of the Xapp targets. Apps which force a specific style on window decorations imo disqualify for that target. (imo same story as styling limitations brought by libadwaita)

If you allow CSD only apps, e.g. Xfce could not stat that "All Xapps will look well integrated in our DE". For the README section `## XApp compliance` we would need to cut out the part `and should integrate well into any desktop environment that has adopted the XApp specification.`, because that won't be the case anymore.

Sure, by allowing CSD-only apps we would have more apps which qualify for beeing an Xapp ... though I think we should rather aim for what we want to accomplish instead of maximizing the number of apps which do fit in.

Reviewer 4: An application can use CSDs and still make them styleable, can they not? I'm not sure that I understand why they have to be mutually exclusive.

Author: > An application can use CSDs and still make them styleable, can they not?

If the user does select a specific decoration in the window manager (e.g. the location and style of the close/maximize/minimize button ... that's currently possibly via xfwm4), then applications which only offer CSD won't be able to follow the selected window style.

> I'm not sure that I understand why they have to be mutually exclusive.

Not mutually exclusive. Fine if an app supports both, CSD and SSD. Though, only supporting CSD imo is bad.

Reviewer 5: That's not entirely true - on GNOME apps use CSD and use the placement settings set by the WM.

Author: > That's not entirely true - on GNOME apps use CSD and use the placement settings set by the WM.

Not sure to what you refer with "That".  Sure, GNOME apps do use CSD and are placed by the WM … as far as I can tell, nobody questioned that.

Reviewer 5: > If the user does select a specific decoration in the window manager (**e.g. the location and style of the close/maximize/minimize button** ... that's currently possibly via xfwm4), **then applications which only offer CSD won't be able to follow the selected window style**.

Author: > > If the user does select a specific decoration in the window manager (**e.g. the location and style of the close/maximize/minimize button** ... that's currently possibly via xfwm4), **then applications which only offer CSD won't be able to follow the selected window style**.

I still don't get your critics.  Let put an example. I use xfwm4 and choose the decorations to look like this: 

![Bildschirmfoto_2024-07-09_23-32-44](https://github.com/xapp-project/xapp-project/assets/2205293/6c068fb9-5dcc-49a3-89a5-54d733bdc8ef)

That works fine for almost all applications I have installed, because they don't force CSD. But there is e.g. GNOME simple-scan, which does force CSD, and as such is not able to follow the window decoration style I selected:

![Bildschirmfoto_2024-07-09_23-33-03](https://github.com/xapp-project/xapp-project/assets/2205293/cd7d3f9b-c704-4bbb-bb15-7b02e9fe3154)

And because of that inability, it looks out of place when used in Xfce.

It might be nice for a single app-developer, DE or toolkit to put foreign decoration, in order to provide an unique look. However, for linux-desktop in general, it is rather bad, since it creates fragmentation and is an annoyance if you want to mix apps cross-desktop.

Reviewer 6: On Wayland (I'd hope?) you can ask the compositor what it wants you to use since many compositors let you choose CSD/SSD globally and sometimes on a per-app basis. CSD tends to decrease accessibility, and I'm sensitive to those issues with Gnome in particular [OT Gnome a11y discrimination rant self-censored 🤬], so if you don't have reason to know the user wants CSD, use SSD is the only reasonable default.

Supporting CSD … Given RH+derivs + Debuntu+derivs pretty much all shipping Gnome default unless the distribution explicitly exists to ship something else … I'd argue "universal" apps probably need to support CSD. If there's no consensus on that, I'd recommend a SHOULD, but I won't protest a MAY too loudly because Gnome is frankly idiotic on this issue … among so many others.

I still use CSD for Firefox because Firefox's UI uses ridiculously too much vertical screen real estate for tabs and browser bar. There just isn't a "compact" theme option. I'd care about that a helluva lot more than I do the pretty themes, but even using the GTK theme doesn't use appreciably less real-estate. Some of that is my font scaling. Some of that is padding around UI elements I really don't care about. It is a great example (even if a singular one) of why a person might prefer CSD in some cases.

Reviewer 3: The intent is not to force users to use SSDs, so perhaps the wording here should be clearer.  What we're trying to get across is: obey the user's preference (either through a hint set by the desktop environment, or through an in-app setting), but supporting CSDs is optional, and lack of CSDs will not prevent an app from being considered an xapp.

Reviewer 3: So perhaps a further revision to my suggested wording:

"Applications which are part of the xapp project should abide by user preferences as to whether server-side or client-side decorations are to be used.  If the windowing system has a protocol for informing apps of the user's preference, it should be honored. However, an xapp is not required to support client-side decorations, and may assume that the windowing system will draw decorations in that case.  An xapp may allow the user to override the windowing system setting, but if it does, by default it must respect the setting advertised by the windowing system."

Reviewer 3: Another possibility, similar, but maybe clearer(?):

"Applications which are part of the xapp project should by default assume that server-side decorations will be drawn.  If the windowing system has a protocol to advertise to applications the user's preference between server- and client-side decorations, the application must honor that preference, with the exception that the application may choose not to support client-side decorations.  Applications may also provide an in-app setting to override this behavior, but the setting must default to honoring the global preference exposed by the windowing system, if any."

Reviewer 6: I thought we were clear that SSD must be supported? My reading of kelnos' wording suggests that to be the case, so I will offer:

"Xapp project applications must support server-side window decorations, where applicable, but may also support client-side decorations. If both are supported, the determination of which to use by default should follow user and environment preferences, falling back to server-side if no default can be determined. Applications may allow users to configure them to use either decoration model."

I feel that wording is still clumsier than I like, but it's the best I can come up with at the moment.

Reviewer 7: Here is a rephrase of @knghtbrd's version.

> Xapp project applications must be capable of utilizing server-side window decorations when applicable. Client-side decorations are also permissible. If both are supported, the default decoration method should prioritize user and environment preferences. If no definitive preference can be established, server-side decorations should be used by default. Applications should provide user-configurable options to select their preferred decoration model.

Author: Phew, too many options :smile_cat:.  In the end I picked the one from @knghtbrd . Found it the most clear and it uses MUST / MAY / SHOULD, which is required for RFC 2119. Thank you all to get that section straight!

### Menus

If menus are required for a xapps, traditional menu bars should be used by default. It is fine to optionally support e.g. hamburger menus [^2].

> TBD: Do we want to force a default on this?

Author: The [original Xapp guide](https://linuxmint-developer-guide.readthedocs.io/en/latest/xapps.html) requested traditional menus, so I added it as well here ... though not sure if we still want that.

Reviewer: I'm not sure I agree with the idea of mandating support for traditional menubars, as this may exclude applications which would not otherwise benefit from them, create unnecessary development / maintenance overhead (if needing to maintain multiple implementations for said functionality), or run counter to any desired HIG they wish to follow (which doesn't necessarily mandate any given theming).

Each application under Xapps should ideally be able to consider the usefulness and/or relevance of the menubar. If the application is written in something like Kirigami + Qt as an example, the application may choose alternative navigation methods such as a toolbar or a navigation toolbar.

- https://develop.kde.org/hig/components/navigation/menubar/
- https://develop.kde.org/hig/components/navigation/toolbar/
- https://develop.kde.org/hig/components/navigation/navigationtabbar/

Applications written in GTK4 (not necessarily with libadwaita) could theoretically use GtkPopoverMenuBar, but given they would be using GtkHeaderBar, it's more likely they would just be putting controls in that unless their application _needs_ more complex navigation (hence they should be considering the usefulness of menubar rather than it being forced).

Reviewer 3: Agreed; I don't think we should mandate menubars be used.  I think they're appropriate in some cases (like OpenOffice, where I use the menus ever time I use the app), but not so much in others (I have Firefox's menubar hidden, and I can't remember the last time I used it).

And similar to my comment about supporting the various standards, I want to make sure that prospective xapp authors don't read this and think that they're required to have a menubar, even if they don't need one.

Author: I don't have a strong opinion on this one, though I tend to agree with @JoshStrobl and @kelnos for it. It would be fine for me to just drop the menubar paragraph ... though let's wait for the opinion of @clefebvre

Reviewer 2: I agree as well. This was written in 2016 at a time where CSD/headerbars shocked people and weren't properly supported.

In many apps nowadays (including existing Xapps) CSD headerbars are used, properly supported and they actually manage to provide a traditional experience.

Author: Alright, dropped the paragraph "Menus": d0503c71f1a4ac227eb3e70da6496d6e716c2874 - closing this thread

### Theming

> User theming has to be supported by xapps (TODO: Where is the spec for general theming?) [^3]

Reviewer 3: I think this will definitely need to be fleshed out more, and maybe should be left out of the initial version of the spec.  It's unclear how an app developer would conform to this requirement.  If they're using something like GTK or Qt, then they get theming built-in.  If they aren't, then it's hard to mandate theming support when there's no general standard for it.

And even when an app is written with a toolkit that supports theming, that theming will be specific to the toolkit, not the desktop, so I'm not sure how we can really mandate anything meaningful here.

Author: Maybe let's put something more general in the section, to make clear that theming generally is welcome / applications should not disable theming support on purpose, if the toolkit allows so.

Suggestion: "If the used toolkit offers the possibility, xapps should be themable"

Reviewer 6: This one seems stalled on not really having specific language to make this more than a vague suggestion. I'd suggest leaving it out for the moment so that it doesn't become a blocker. I do want this in the spec, but with wayland adoption accelerating rather quickly rather suddenly, I'd be somewhat hesitant to recommend a mechanism for applications to automatically pick up even your vague presentation preferences without creating an xkcd 927 sort of situation.

This is a great example of what I meant in my issue about it being easier to discuss things in discrete issues or PRs once we have a generally accepted spec to hang them on, even if just a draft.

### Desktop Environment Agnostic

> If the same windowing, theme and icon theme is used, XApps are supposed to look the same on each Linux desktop environment they run on. As such, they should not have dependencies to desktop-specific styles.

Reviewer: ```suggestion
If the same windowing, theme and icon theme is used, XApps are intended to be visually consistent across Linux desktop environments. As such, they should not have dependencies to desktop-specific styles.
```

### Identification as xapp application

If an application fulfills all above criteria and is supposed to be added to the xapp project, it has to add a paragraph to a README.md file located on the top level of its git project with the following text:

> ## Xapp compliance
>

> > This application follows the standards defined in the [Xapp specification v1.0.0](insert link here) and as such will run fine on linux desktop environments which support freedesktop.org standards. The UI of this application will be well integrated on desktop environments which support the xapp standard.

Reviewer 3: "will run fine" sounds kinda weird to me, and maybe leave out "Linux".  And calling out fd.o is maybe a little too restrictive; if we adopt/mandate standards outside of the fd.o umbrella, this will become out of date.  Overall, I think this required statement is a little too verbose.  I'm thinking of something simpler and shorter, like:

"This applications follows the standards defined in version x.y.z of the Xapp specification, and should integrate well into any desktop environment that has adopted the Xapp specification."

Author: Indeed, better to be more generic here, let's just use your suggestion: a852adcac09a3c86de2bb10c8e241a5add00f118

> The xapp project provides a set of [project-properties](https://docs.github.com/en/enterprise-cloud@latest/organizations/managing-organization-settings/managing-custom-properties-for-repositories-in-your-organization) which allows to filter xapp projects for supported operating systems, the used license, the used toolkit and supported display servers. Please set these project-properties for your project accordingly.  

Author: ```suggestion
The xapp project provides a set of [project-properties](https://docs.github.com/en/enterprise-cloud@latest/organizations/managing-organization-settings/managing-custom-properties-for-repositories-in-your-organization) which allows to filter xapps for supported operating systems, the used license, the used toolkit and supported display servers. Please set these project-properties for your project accordingly.  
```

Improved wording

### Repository Location

> Once an application is approved as a xapp application, it either is hosted directly by the Xapp project (so that the native issue tracker, git repository and infrastructure is used) or its repository is mirrored to the xapp project. On choosing the latter, the xapp project page will hide the possibility to file issues and show a link to the original repository location.

Reviewer: ```suggestion
Once an application is approved as an xapp application, it is either hosted directly by the Xapp project (so that the native issue tracker, git repository and infrastructure is used) or its repository is mirrored to the xapp project. On choosing the latter, the xapp project page will hide the possibility to file issues and show a link to the original repository location.
```

> Once an application is approved as an xapp application, it is either hosted directly by the Xapp project (so that the native issue tracker, git repository and infrastructure is used) or its repository is mirrored to the xapp project. On choosing the latter, the xapp project page will hide the possibility to file issues and show a link to the original repository location.

Reviewer 3: This feels a little ambitious; what's the rationale behind encouraging projects to use Xapp hosting (where is that?), or even mirroring it?  Wouldn't Xapp just hosting a list of conformant projects be sufficient, and low overhead for all concerned?

Reviewer 8: Ideally, there should be a differentiation between apps that will be (or want to be) developed under the xapp-project umbrella, and other apps that aren't necessarily under the xapp-project umbrella but do follow the XApp standards which we say are xapp-compliant applications (**all projects being developed under the xapp-projects umbrella must be xapp-compliant but not all xapp-compliant applications must be developed under the xapp-projects umbrella, they just must have to follow the standards**). I believe we should clarify this differentiation and that the repository location requirements only apply to applications that want to be developed/supported/maintained under the xapp-projects umbrella.

Author: True, a filterable list of project names and links which do fulfill the xapp spec would as well be sufficient for the spec alone.

Afaik the idea of @clefebvre was to host some existing linux mint xapps outside of the linux mint project, to make clear that they are supposed to be used by other DE's as well .. in addition, there are plans to fork some of the gnome apps, like e.g. "gnome-scan", revert them to some previous state and maintain them here as a shared effort between DE's supporting xapps. (Here the [related Mint Blogpost](https://blog.linuxmint.com/?p=4675))

Foreign repositories hosting xapps (e.g. several Xfce applications) could then be mirrored to the xapp-project in order to be shown alongside the native ones, in order to get the same attention as these (I suppose that would be the only advantage of mirroring over just maintaining a list)

Reviewer 2: We'll need a list no matter what, at least on the website. And if it's easy to automate github mirroring let's do it.

I think different people look at different things. Casual users might pay more attention to a website, devs might not even look at it but they might browse the list of repositories on xapp-project.

Being present in both places gives more visibility to the app and makes it clearer it's on board with the project.

Reviewer 2: If for any reason we need two different levels we can show that in the list, and with badges on the READMEs, once it's mirrored it's visible anyway, it shouldn't really matter whether an app is primarily hosted on the XApp github or somewhere else imo.

Author: Just played a bit with Github custom project properties, I suppose they would be superior to badges on the README, because like that, it would be possible to filter for apps which fulfill specific criteria. [Here a minimal filter example](https://github.com/orgs/test-alexxcons-organization/repositories?q=props.Windowing%3AWayland) (just write `prop`, and you will get an auto-complete). [Here all props of the project](https://github.com/test-alexxcons-organization/testrepo/settings/custom-properties).

E.g. if I want to know which XApps do support Wayland, having a possibility to filter will get important.

However, for sure we could as well implement a filter on the website ... maybe both?
